### PR TITLE
Improved mock middleware

### LINKED ISSUE #130: Bad response when filename is too long with mock record mode
I have an issue with record mode when the url is very long. I'm on mac and the limit for filename is 255 chars.

The filename is :
$ echo -n "GET____aaa-bbbbbb.search.windows.net_indexes-clips-index-docs-search=--facet-0-=createdAt-interval-day--skip=0--top=3--orderby=createdAt-desc-api-version=2015-02-28--filter=-userId-eq--c65c6f24-33a6-4b3b-b557-aaaaaaaaaaaa--____7e8914e63a4de9a74132056db7b14026.txt" | wc -c
     263

The end of filename is a [md5()](https://github.com/csarrazi/CsaGuzzleBundle/blob/d2dff62f436d14ab0321ef8a558790bd5161be60/src/GuzzleHttp/Cache/MockStorageAdapter.php#L112) I think it's possible to reduce it size to 7 or 8 chars with substr().

### PR DESCRIPTION
**Pull request by csarrazi**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #130
| License       | MIT

Refactored naming strategies for mock and cache middleware, and added support for subfolder naming strategy. Should fix #130.

### CODE REVIEW

**Comment by iamluc** [comment]
The refactoring LGTM.

Also, something I was thinking about was to add a map to help finding the mock file.

ie. in `mockmap.json`
```json
{
    "hosts": {
        "_default_": {
            "/app_dev.php/my_page?param1=test": "GET____app_dev_php_my_page_param1_xxxxx.txt"
        }
    }
}
```

**Comment by SofHad** [comment]
+1

**Comment by csarrazi** [comment]
@iamluc Unfortunately, it's not as simple. The fingerprint is generated not only from request query params and the path, but also the headers. Because of this, this won't work with the structure you recommended.

File: src/GuzzleHttp/Cache/MockStorageAdapter.php [hunk 1]
```diff
@@ -83,59 +69,26 @@ public function save(RequestInterface $request, ResponseInterface $response)
...
-        } else {
-            $path = sprintf(
-                '%s_%s____%s',
-                str_pad($request->getMethod(), 6, '_'),
-                urldecode(ltrim($request->getUri()->getPath(), '/').'-'.$request->getUri()->getQuery()),
-                $fingerprint
-            );
-        }
-
-        return $this->storagePath.'/'.preg_replace('/[^a-zA-Z0-9_+=@\-\?\.]/', '-', $path).'.txt';
```
Comment (SofHad) [inline src/GuzzleHttp/Cache/MockStorageAdapter.php]: I suggest to dispatch an event (Symfony EventDispatcher) here in order to allow users, with a listener, to have a custom storage path based on the request ?
For example:
{ storagePath }/api1/GET/hash.txt
{ storagePath }/api1/POST/hash.txt
{ storagePath }/api2/GET/hash.txt
....

Comment (csarrazi) [inline src/GuzzleHttp/Cache/MockStorageAdapter.php]: You can already do this by creating a new instance of the mock middleware and mock storage adapter, with their own configuration.

Also, I do not think that this is a valid use case for the event dispatcher.

**Pull request merged**